File: backend/src/main/java/br/com/cartinhas/entity/Deck.java
package br.com.cartinhas.entity;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.OrderBy;
import javax.persistence.Transient;

import br.com.cartinhas.enuns.ERarity;
import org.hibernate.annotations.LazyCollection;
import org.hibernate.annotations.LazyCollectionOption;
import org.springframework.beans.factory.annotation.Autowired;

import br.com.cartinhas.entity.converter.CardConverter;
import br.com.cartinhas.entity.dto.DeckDTO;
import br.com.cartinhas.enuns.EColor;

@Entity
public class Deck {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    private String name;

    @OrderBy("idReference")
    @LazyCollection(LazyCollectionOption.FALSE)
    @ManyToMany()
    @JoinTable(name = "DECK_CARTAS")
    private List<Card> cards;

    @Transient
    private String cardImp;

    @Transient
    private Long countM;

    @Transient
    private Long countR;

    @Transient
    private Long countU;

    @Transient
    private Long countC;
    
    @Autowired
    @Transient
    private CardConverter cardConverter;

//    Constructor
    public Deck() { }
    
    public Deck(DeckDTO dto) {
    	this.id = dto.getId();
		this.name = dto.getName();
		this.cards = cardConverter.convertToEntity(dto.getCards());
		this.countM = dto.getCountM();
		this.countR = dto.getCountR();
		this.countU = dto.getCountU();
		this.countC = dto.getCountC();
    }

    public Deck(String name) {
        this.name = name;
    }

//    Getters and Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public List<Card> getCards() {
        if(cards ==null){
            cards = new ArrayList<>();
        }
        return cards;
    }

    public void setCards(List<Card> cards) {
        this.cards = cards;
    }

    public String getCardImp() {
        return cardImp;
    }

    public void setCardImp(String cardImp) {
        this.cardImp = cardImp;
    }

    public List<EColor> getColors() {
    	
    	List<EColor> colors = new ArrayList<EColor>();
    	cards.forEach(c -> colors.addAll(c.getColors()));
        return colors.stream().distinct().collect(Collectors.toList());
    }

    public Long getCountM() {
        return cards.stream().filter(c -> c.getRarity().equals(ERarity.MYTHIC_RARE)).count();
    }

    public Long getCountR() {
        return cards.stream().filter(c -> c.getRarity().equals(ERarity.RARE)).count();
    }

    public Long getCountU() {
        return cards.stream().filter(c -> c.getRarity().equals(ERarity.UNCOMMON)).count();
    }

    public Long getCountC() {
        return cards.stream().filter(c -> c.getRarity().equals(ERarity.COMMON)).count();
    }
}

File: frontend/src/app/deck/models/deck.ts
import { EnumTypes } from './../../enums/e-types';
import { Card } from './card';
import { Color } from './color';

export class Deck {
    public static types = Object.keys(EnumTypes);

    private _id: number;
    private _name: string;
    private description: string;
    private _colors: Array<Color>;
    private _cards: Array<Card>;
    private countM: number;
    private countR: number;
    private countU: number;
    private countC: number;
    private _cardImp: string;

    public get id() {
        return this._id;
    }

    public set id(id: number) {
        this._id = id;
    }

    public get name() {
        return this._name;
    }

    public set name(name: string) {
        this._name = name;
    }

    public get colors() {
        return this._colors;
    }

    public get cardImp() {
        return this._cardImp;
    }

    public set cardImp(cardImp: string) {
        this._cardImp = cardImp;
    }

    public get cards() {
        return this._cards;
    }

    public set cards(cards) {
        this._cards = cards;
    }

    constructor(obj?) {
        return Object.assign(this, obj);
    }

    allTypes() {
        return Object.values(EnumTypes);
    }

    list(type: EnumTypes) {
        if (this.cards) {
            const cards = this.cards.filter(e => e.types[0] === type && this.cards.find(e2 => e2.name === e.name) === e);
            return cards.length > 0 ? cards : false;
        }
    }

    countTypes(type: EnumTypes) {
        if (this.cards) {
            return this.cards.filter(e => e.types[0] === type).length;
        }
    }

    get deckImage() {
        return this.cards.find(e => e.imageUrl != null && this.contCard(e) > 0).imageUrl;
    }

    contCard(card: Card) {
        return this.cards.reduce((prevVal, elem) => {
            return elem.name === card.name ? prevVal += 1 : prevVal;
        }, 0);
    }
}
